<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN">
<html>
<head>
<title>LBIT</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<!-- Import scripts in this specific order -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.1/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jqueryui/1.10.4/jquery-ui.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/qtip2/2.2.0/jquery.qtip.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/toastr.js/2.0.2/js/toastr.js"></script>
<script src="webplugin/ete.js"></script>
<script>
/* YOU MUST SET THIS VARIABLE */
var ete_webplugin_URL = "cgi-bin/webplugin.py"

</script>
<link rel="stylesheet" type="text/css" href="webplugin/ete.css">
<link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/qtip2/2.2.0/jquery.qtip.min.css">
<link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/toastr.js/2.0.2/css/toastr.min.css">
</head>

<body>
<i>Menu</i>  <b>::</b> <a href="index.html">Draw Tree</a><br><br>

<!-- This layer must be called "popup" - Used to display the menu when clicking a node -->
<div id="popup"></div>

<div id="content" style='text-align: left;'>
    <ul>
        <h2>Species Tree</h2>
        <li>
        <textarea
            title='• You can paste trees in newick (NH) and extended newick (NHX) formats. <br><br>
                   • Features in NHX format will be parsed as tree features so you can add them as node faces to the tree image.<br><br>
                   • Features whose names starts by "_" will be ignored. <br><br>
                   • Case sensitive.'
            id="speciesTree" onClick="$(this).removeClass('disabled');$('#sp_tol_form').addClass('disabled');$('#sp_tol_form').find('#sp_tol').prop('checked', false);"></textarea>

        <br><i> - OR - </i><br>
        <form id="sp_tol_form" class="disabled">
            <label for="sp_tol" style="display:inline-block" >
                Ensembl Species Tree
                <input id="sp_tol" title="Ensembl tree has lower case species names" type="checkbox" name="sp_tol" onClick="$('#sp_tol_form').toggleClass('disabled');$('#speciesTree').toggleClass('disabled');" />
            </label>
        </form>
        </li>

        <h2>Gene Tree</h2>
        <li>
        <textarea
            title='• Labels should follow the format "GENE;;SPECIE". <br><br>
                   • Case sensitive.'
            id="geneTree" onClick="$(this).removeClass('disabled');$('#gn_ensembl_form').addClass('disabled');"></textarea>
        <br><i> - OR - </i><br>
        <form id="gn_ensembl_form" class="disabled" onClick="$(this).removeClass('disabled');$(this).find('#gn_ensembl').focus();$('#geneTree').addClass('disabled');">
            Ensembl Gene Tree ID : <input id="gn_ensembl" type="textbox" name="gn_ensembl" >
        </form>
        </li>

        <h2>Gene Distances</h2>
        <li><textarea title="Distance matrix in PHYLIP format (full or lower triangular)" id="geneDistances"></textarea></li>

        <h2>Gene Sequences</h2>
        <li><textarea title="FASTA, NEXUS or PHYLIP format" id="geneSeq"></textarea><br>
          Format :
          <select id="seq_format">
              <option value="nexus">Nexus</option>
              <option value="fasta">Fasta</option>
              <option value="phylip">Phylip</option>
          </select><br>
          Data type :
          <select id="seq_data_type">
              <option value="dna">DNA</option>
              <option value="rna">RNA</option>
              <option value="protein">Protein</option>
          </select><br>
            <label for="seq_align" title="ClustalO v1.2.0" style="display:inline-block" >
                Align
                <input id="seq_align" type="checkbox" name="seq_align" />
            </label> <br>
            <label for="seq_calculate_dm" title="ClustalO v1.2.0" style="display:inline-block" >
               Calculate distance matrix from sequences
                <input id="seq_calculate_dm" type="checkbox" name="seq_align" />
            </label>

        </li>

        <h2>PolytomySolver</h2>
        <li>
          <label for="solve_polytomy" style="display:inline-block" >
          Run PolytomySolver
                <input id="solve_polytomy" type="checkbox" name="solve_polytomy" />
          </label> <br>
          Mode :
          <select id="gn_reroot_mode">
              <option value="outputallroots">Reroot (outputallroots)</option>
              <option value="none">Use given root (none)</option>
              <option value="findbestroot">Reroot and return best solution (findbestroot)</option>
          </select>
          <br>
            <label for="gn_contract_branches" style="display:inline-block">
            <input id="gn_contract_branches" type="checkbox" name="seq_align" />
            Contract gene tree branches with support lower than : <input id="gn_support_threshold" type="textbox" value="0.0" name="gn_support_threshold">
            </label>
          <br>
            <label for="poly_sol_limit" style="display:inline-block">
            Solution Limit : <input id="poly_sol_limit" type="textbox" value="20" >
            </label>
          <br>
            <label for="poly_path_limit" style="display:inline-block">
            Path Limit : <input id="poly_path_limit" type="textbox" value="5" >
            </label>
        </li>
        <h2>ParalogyCorrector</h2>
        <li>
          <label for="correct_paralogy" style="display:inline-block" >
          Run ParalogyCorrector
                <input id="correct_paralogy" type="checkbox" name="correct_paralogy" />
          </label> <br>
                Orthologs :
                <input id="add_ortholog" type='button' value='Add'>
                <input id="rm_ortholog" type='button' value='Remove'>
                <input id="rm_all_orthologs" type='button' value='Remove All'>
            <div id="orthologs">
            </div>
        </li>
    </ul>
</div>

<div style="clear:both;" >
    <input type='submit'  value='Draw tree'  onClick='polytomysolver(random_tid(), $("#speciesTree").val(),$("#geneTree").val() ,$("#geneDistances").val(), $("#geneSeq").val(),"name", $("#sp_tol:checked").length,$("#gn_ensembl").val(),$("#gn_reroot_mode option:selected").val(),$("#gn_support_threshold").val(),$("#gn_contract_branches:checked").length,$("#seq_format option:selected").val(),$("#seq_align:checked").length,$("#seq_data_type option:selected").val(),$("#seq_calculate_dm:checked").length,$("#poly_sol_limit").val(),$("#poly_path_limit").val(), get_orthologs(), $("#correct_paralogy:checked").length, $("#solve_polytomy:checked").length,"#polytomysolver_dropdown");'>


    <input type='button' value='Clear' onClick='$(".ete_image").html("");'>
</div>
    </div>
    <br>
    <!-- Tree containers -->
    <div class="ete_image_container" style="display:inline-block;" id="img0">
    	<div class="ete_image" style="display:inline-block;" id="polytomysolver_dropdown"></div><br>
    	<div class="ete_image" style="display:inline-block;" id="img1"></div><br>
	</div>

    </body>
</html>
